### PR TITLE
feat: add `ApprovalRevocationEnforcer`

### PR DESCRIPTION
**Pull request by jeffsmale90**
### **What?**

This enforcer grants the authority to revoke allowances granted by either:

- ERC20 `approve(spender,amount)`
- ERC721 `approve(to,tokenId)`
- ERC721 `setApprovalForAll(operator,approved)`
- Permit2 `approve(token, spender, 0, 0)`
- Permit2 `lockdown((address,address)[])`
- Permit2 `invalidateNonces(token, spender, newNonce)`

This covers approvals set for ERC20, ERC721, ERC1155 tokens, and permit2 approvals (both onchain, and offchain).

### **Why?**

For ERC20, and ERC721 approvals, the enforcer now verifies that the target implements the expected token standard, by first invoking the standard-specific function to check the allowance being revoked. For Permit2 approvals, the execution must target the canonical Permit2 address.

Advanced Permissions currently has an `erc20-token-revocation` permission that grants the authority to revoke only ERC20 approvals. NFT approvals are also required.

By combining these revocations into a single enforcer, we get a number of benefits:

- User must sign only a single permission to revoke _all_ allowances across multiple token types
- Reduced gas cost to invoke a revocation (single caveat, rather than composition of multiple caveats)

### **How?**

The enforcer accepts terms of exactly 1 byte, interpreted as a bitmask of the following values:

- Bit 0 (`0x01`) - ERC-20 approve(spender, 0) (spender non-zero, amount zero)
- Bit 1 (`0x02`) - ERC-721 per-token approve(address(0), tokenId)
- Bit 2 (`0x04`) - ERC-721 / ERC-1155 setApprovalForAll(operator, false)
- Bit 3 (`0x08`) - Permit2 `approve(token, spender, 0, 0)` against the canonical Permit2 deployment
- Bit 4 (`0x10`) - Permit2 `lockdown((address,address)[])` against the canonical Permit2 deployment
- Bit 5 (`0x20`) - Permit2 `invalidateNonces(token, spender, newNonce)` against the canonical Permit2 deployment

Indicating which revocation primitives the delegation authorizes. Terms must be non-zero, and the reserved upper bits must not be set.

The beforeHook only runs in single call type and default execution mode. It first performs general verification - no native value is sent (no additional value limiting caveat is required), and the calldata carries at least a 4-byte selector - then dispatches by selector.

Permit2 primitives are dispatched first, since each has its own unique selector and a target check against the canonical Permit2 deployment. `approve(token, spender, uint160, uint48)` additionally requires the calldata to be exactly 132 bytes with both the amount and expiration parameters zeroed. `lockdown((address,address)[])` and `invalidateNonces(address,address,uint48)` have otherwise unconstrained calldata - the Permit2 contract itself structurally guarantees they can only zero per-pair allowance amounts or strictly increment nonces, never grant authority.

If the selector is not a Permit2 one, the calldata must be exactly 68 bytes (4-byte selector + two 32-byte words). `setApprovalForAll` and `approve` are then distinguished by selector. The two `approve` signatures share a selector and are distinguished by the `spender` / `to` (first) parameter — if it is the zero address, the call is treated as the ERC-721 `approve(to, tokenId)` form; otherwise it is treated as the ERC-20 `approve(spender, amount)` form. Any other selector is rejected.

The ERC-20, ERC-721, and `setApprovalForAll` branches then perform a check against the delegator's current approval state on the target, ensuring the revocation is to an existing approval - ensuring that the contract implementation is a valid target for the invocation. The Permit2 branches perform no such liveness pre-check; their structural calldata constraints already guarantee the call cannot grant authority, so an absent on-chain allowance is a harmless no-op (or, for `invalidateNonces`, reverts inside Permit2 itself).

---

> [!NOTE]
> **High Risk**
> Adds a new caveat enforcer that gates token approval/Permit2 revocations via calldata inspection and on-chain checks, touching permissioning behavior around asset approvals. Mistakes could allow unintended calls or create denial-of-service scenarios (notably via Permit2 `invalidateNonces`) when used in delegations.
> 
> **Overview**
> Adds a new `ApprovalRevocationEnforcer` caveat that allows delegates to *only revoke* existing approvals (ERC-20 `approve(spender,0)`, ERC-721 `approve(0,tokenId)`, `setApprovalForAll(false)`, and three canonical Permit2 revocation primitives) controlled by a 1-byte `terms` bitmask and restricted to single/default execution.
> 
> Updates deployment/verification scripts to include the new enforcer, adds extensive unit/integration tests (including redelegation edge cases), and documents the enforcer’s intended usage, trust assumptions (canonical Permit2 address), and known incompatibilities (ERC-20s that revert on zero-amount `approve`).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit cfa5bcb77de6001d70ab96c45588666216653d3d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: script/verification/verify-enforcer-contracts.sh [hunk 1]
```diff
@@ -94,6 +95,7 @@ ADDRESSES=(
   "0x44877cDAFC0d529ab144bb6B0e202eE377C90229"
   "0x9eB86bbdaA71D4D8d5Fb1B8A9457F04D3344797b"
   "0xaD551E9b971C1b0c02c577bFfCFAA20b81777276"
+  "0x0000000000000000000000000000000000000000"
```
[the remark this answers is not in the document]
Comment (MoMannn) [reply script/verification/verify-enforcer-contracts.sh:98]: intended, will be added at deploy time.

File: src/enforcers/ApprovalRevocationEnforcer.sol [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+        (address target_, uint256 value_, bytes calldata callData_) = _executionCallData.decodeSingle();
+
+        require(value_ == 0, "ApprovalRevocationEnforcer:invalid-value");
+        // 68 = 4-byte selector + two 32-byte words. Shared by `approve(address,uint256)` and
+        // `setApprovalForAll(address,bool)`.
+        require(callData_.length == 68, "ApprovalRevocationEnforcer:invalid-execution-length");
+
+        if (bytes4(callData_[0:4]) == IERC721.setApprovalForAll.selector) {
+            require(flags_ & _PERMISSION_SET_APPROVAL_FOR_ALL != 0, "ApprovalRevocationEnforcer:permission-not-granted");
```
Comment (jeffsmale90) [inline src/enforcers/ApprovalRevocationEnforcer.sol]: `permission-not-granted` might be confusing here as we don't indicate what permission is not granted.

Maybe something like `revert-method-not-allowed`, or `revocation-mechanism-not-allowed` maybe more clear?

We could be even more clear with `setApprovalForall-approval-method-not-allowed` or something.

File: src/enforcers/ApprovalRevocationEnforcer.sol [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+    ////////////////////////////// Internal Methods //////////////////////////////
+
+    /**
+     * @dev Validates and returns the raw permission flags byte. Reverts on invalid terms.
+     */
+    function _parseFlags(bytes calldata _terms) private pure returns (uint8 flags_) {
+        require(_terms.length == 1, "ApprovalRevocationEnforcer:invalid-terms-length");
+        flags_ = uint8(_terms[0]);
+        require(flags_ != 0, "ApprovalRevocationEnforcer:no-permissions");
```
Comment (jeffsmale90) [inline src/enforcers/ApprovalRevocationEnforcer.sol]: similar to my other comment - I'm not sure `permissions` is the best term here, as it's ambiguous.

**Review by jeffsmale90 — COMMENTED** [review]
Looks good to me!

Minor comment regarding using the term `permission` to refer to the allowed revocation mechanism.

File: src/enforcers/ApprovalRevocationEnforcer.sol [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        uint8 flags_ = _parseFlags(_terms);
+
+        (address target_, uint256 value_, bytes calldata callData_) = _executionCallData.decodeSingle();
+
+        require(value_ == 0, "ApprovalRevocationEnforcer:invalid-value");
+        // 68 = 4-byte selector + two 32-byte words. Shared by `approve(address,uint256)` and
+        // `setApprovalForAll(address,bool)`.
+        require(callData_.length == 68, "ApprovalRevocationEnforcer:invalid-execution-length");
+
+        if (bytes4(callData_[0:4]) == IERC721.setApprovalForAll.selector) {
```
Comment (McOso) [inline src/enforcers/ApprovalRevocationEnforcer.sol]: one readability improvement and perhaps minor gas improvement is caching selector, first arg, and second arg before if logic.

Something like:
```
bytes4 selector_ = bytes4(callData_[0:4]);
address firstArg_ = address(bytes20(callData_[16:36]));
uint256 secondArg_ = uint256(bytes32(callData_[36:68]));
```

**Review by McOso — DISMISSED** [review]

**Review by jeffsmale90 — COMMENTED** [review]
This looks good to me!

**Review by MoMannn — APPROVED** [review]

**Review by McOso — APPROVED** [review]
